README.txt
About Ed25519
=============

Ed25519 is a public-key signature system with several attractive features 
including:

* Fast single-signature verification.
* Very fast signing.
* Fast key generation.
* High security level.
* Small signatures. Signatures fit into 64 bytes.
* Small keys. Public keys consume only 32 bytes. 

This text abridged from http://ed25519.cr.yp.to/.

About ed25519ll
===============

ed25519ll is a low-level cffi wrapper for the Ed25519 public key signature
system. It uses platformer to compile a shared library that is not a
Python extension module, and then uses cffi to talk to the library.

This wrapper currently exposes the supercop-ref10 implementation of
Ed25519, on my 2.6GHz Athlon achieving about 7200 signatures/second/core
and 2900 verifications/second/core including the wrapper overhead.

As of version 0.3, ed25519ll cannot be built automatically by pip, as the 
ed25519 shared library build script (included) is not integrated into setup.py

Example::
    
    import ed25519ll
    msg = b"The rain in Spain stays mainly on the plain"
    kp = ed25519ll.crypto_sign_keypair()
    signed = ed25519ll.crypto_sign(msg, kp.sk) 
    verified = ed25519ll.crypto_sign_open(signed, kp.vk)
    assert verified == msg  # but ValueError is raised for bad signatures 



CHANGES.txt
0.4
===
- Switch away from ffi.verify() again

0.3
===
- Switch to supercop-ref10
- Switch to ffi.verify()
- Update to work with cffi 0.3
- Update to work with Python 3.3 

0.2
===
- 100% test coverage
- Basic documentation

0.1
===

- Initial release


ed25519ll/test.py
# -*- coding: utf-8 -*-

from nose.tools import raises
import ed25519ll 

def test_ed25519ll():
    msg = b"The rain in Spain stays mainly on the plain"
    kp = ed25519ll.crypto_sign_keypair()
    signed = ed25519ll.crypto_sign(msg, kp.sk)    
    
    @raises(ValueError)
    def open_corrupt(): 
        corrupt = signed[ed25519ll.SIGNATUREBYTES:] + b'U' + signed[ed25519ll.SIGNATUREBYTES+1:]
        ed25519ll.crypto_sign_open(corrupt, kp.vk)                
    open_corrupt()
    
    @raises(AssertionError)
    def short_sk():
        signed = ed25519ll.crypto_sign(msg, kp.sk[:-1])
    short_sk()
    
    ed25519ll.crypto_sign_open(signed, kp.vk)
    
def test_cover_warn_seed():
    ed25519ll.crypto_sign_keypair('*'*32)
    
@raises(ValueError)
def test_bad_seed_size():
    ed25519ll.crypto_sign_keypair('*'*31)

if __name__ == "__main__":
    test_ed25519ll()


ed25519ll/_ed25519.py
import pkg_resources
import sysconfig
import os.path
import glob

from distutils.util import get_platform
from cffi import FFI
from distutils.extension import Extension
ffi = FFI()

decl = """
    extern int crypto_sign(unsigned char *, unsigned long long *,
        const unsigned char *, unsigned long long, const unsigned char *);
    
    extern int crypto_sign_open(unsigned char *, unsigned long long *, 
        const unsigned char *, unsigned long long, const unsigned char *);
    
    extern int crypto_sign_keypair(unsigned char *pk, unsigned char *sk, 
        unsigned char *seed);
"""

ffi.cdef(decl)

verify = True

plat_name = get_platform().replace('-', '_')
so_suffix = sysconfig.get_config_var('SO')
library_dir = os.path.abspath(pkg_resources.resource_filename('ed25519ll', ''))
if not verify: # pragma no cover
    lib_filename = pkg_resources.resource_filename('ed25519ll', 
                                                   '_ed25519_%s%s' %
                                                   (plat_name, so_suffix))
    _ed25519 = ffi.dlopen(os.path.abspath(lib_filename))
else:
    so_prefix = ""
    if os.name == "posix":
        so_prefix = ":" # to link with libraries without 'lib' prefix
    srcdir = os.path.join('..', 'ed25519-supercop-ref10')
    sources  = glob.glob(os.path.join(srcdir, '*.c'))
    include_dirs = [srcdir]
    _ed25519 = ffi.verify(decl,
                          export_symbols=["crypto_sign",
                                          "crypto_sign_open",
                                          "crypto_sign_keypair"],
                          include_dirs=[srcdir],
                          sources=sources,
                          )

crypto_sign_keypair = _ed25519.crypto_sign_keypair
crypto_sign = _ed25519.crypto_sign
crypto_sign_open = _ed25519.crypto_sign_open


ed25519-supercop-ref/ed25519.c
#include "crypto_sign.h"

#include "crypto_verify_32.h"
#include "sha512.h"

#include "ge25519.h"

static void get_hram(unsigned char *hram, const unsigned char *sm, const unsigned char *pk, unsigned char *playground, unsigned long long smlen)
{
  unsigned long long i;

  for (i =  0;i < 32;++i)    playground[i] = sm[i];
  for (i = 32;i < 64;++i)    playground[i] = pk[i-32];
  for (i = 64;i < smlen;++i) playground[i] = sm[i];

  crypto_hash_sha512(hram,playground,smlen);
}


int crypto_sign_keypair(
    unsigned char *pk,
    unsigned char *sk,
	unsigned char *seed
    )
{
  sc25519 scsk;
  ge25519 gepk;
  unsigned char extsk[64];
  int i;

  crypto_hash_sha512(extsk, seed, 32);
  extsk[0] &= 248;
  extsk[31] &= 127;
  extsk[31] |= 64;

  sc25519_from32bytes(&scsk,extsk);
  
  ge25519_scalarmult_base(&gepk, &scsk);
  ge25519_pack(pk, &gepk);
  for(i=0;i<32;i++)
    sk[32 + i] = pk[i];
  return 0;
}

int crypto_sign(
    unsigned char *sm,unsigned long long *smlen,
    const unsigned char *m,unsigned long long mlen,
    const unsigned char *sk
    )
{
  sc25519 sck, scs, scsk;
  ge25519 ger;
  unsigned char r[32];
  unsigned char s[32];
  unsigned char extsk[64];
  unsigned long long i;
  unsigned char hmg[crypto_hash_sha512_BYTES];
  unsigned char hram[crypto_hash_sha512_BYTES];

  crypto_hash_sha512(extsk, sk, 32);
  extsk[0] &= 248;
  extsk[31] &= 127;
  extsk[31] |= 64;

  *smlen = mlen+64;
  for(i=0;i<mlen;i++)
    sm[64 + i] = m[i];
  for(i=0;i<32;i++)
    sm[32 + i] = extsk[32+i];

  crypto_hash_sha512(hmg, sm+32, mlen+32); /* Generate k as h(extsk[32],...,extsk[63],m) */

  /* Computation of R */
  sc25519_from64bytes(&sck, hmg);
  ge25519_scalarmult_base(&ger, &sck);
  ge25519_pack(r, &ger);
  
  /* Computation of s */
  for(i=0;i<32;i++)
    sm[i] = r[i];

  get_hram(hram, sm, sk+32, sm, mlen+64);

  sc25519_from64bytes(&scs, hram);
  sc25519_from32bytes(&scsk, extsk);
  sc25519_mul(&scs, &scs, &scsk);
  
  sc25519_add(&scs, &scs, &sck);

  sc25519_to32bytes(s,&scs); /* cat s */
  for(i=0;i<32;i++)
    sm[32 + i] = s[i]; 

  return 0;
}

int crypto_sign_open(
    unsigned char *m,unsigned long long *mlen,
    const unsigned char *sm,unsigned long long smlen,
    const unsigned char *pk
    )
{
  int i, ret;
  unsigned char t2[32];
  ge25519 get1, get2;
  sc25519 schram, scs;
  unsigned char hram[crypto_hash_sha512_BYTES];

  *mlen = (unsigned long long) -1;
  if (smlen < 64) return -1;

  if (ge25519_unpackneg_vartime(&get1, pk)) return -1;

  get_hram(hram,sm,pk,m,smlen);

  sc25519_from64bytes(&schram, hram);

  sc25519_from32bytes(&scs, sm+32);

  ge25519_double_scalarmult_vartime(&get2, &get1, &schram, &ge25519_base, &scs);
  ge25519_pack(t2, &get2);

  ret = crypto_verify_32(sm, t2);

  if (!ret)
  {
    for(i=0;i<smlen-64;i++)
      m[i] = sm[i + 64];
    *mlen = smlen-64;
  }
  else
  {
    for(i=0;i<smlen-64;i++)
      m[i] = 0;
  }
  return ret;
}


ed25519ll/__init__.py
# -*- coding: utf-8 -*-

import warnings
import os
import os.path
import pkg_resources
from collections import namedtuple
from distutils import sysconfig
from distutils.util import get_platform
import ctypes

__all__ = ['crypto_sign', 'crypto_sign_open', 'crypto_sign_keypair', 'Keypair',
           'PUBLICKEYBYTES', 'SECRETKEYBYTES', 'SIGNATUREBYTES']

plat_name = get_platform().replace('-', '_')
so_suffix = sysconfig.get_config_var('SO')
lib_filename = pkg_resources.resource_filename('ed25519ll', 
                                               '_ed25519_%s%s' %
                                               (plat_name, so_suffix))

_ed25519 = ctypes.cdll.LoadLibrary(os.path.abspath(lib_filename))

PUBLICKEYBYTES=32
SECRETKEYBYTES=64
SIGNATUREBYTES=64
    
Keypair = namedtuple('Keypair', ('vk', 'sk')) # verifying key, secret key

def crypto_sign_keypair(seed=None):
    """Return (verifying, secret) key from a given seed, or os.urandom(32)"""
    vk = ctypes.create_string_buffer(PUBLICKEYBYTES)
    sk = ctypes.create_string_buffer(SECRETKEYBYTES)
    if seed is None:
        seed = os.urandom(PUBLICKEYBYTES)
    else:
        warnings.warn("ed25519ll should choose random seed.",
                      RuntimeWarning)
    if len(seed) != 32:
        raise ValueError("seed must be 32 random bytes or None.")
    s = ctypes.c_char_p(seed)
    rc = _ed25519.crypto_sign_keypair(vk, sk, s)
    if rc != 0: # pragma no cover (no other return statement in C)
        raise ValueError("rc != 0", rc)
    return Keypair(vk.value, sk.value)


def crypto_sign(msg, sk):
    """Return signature+message given message and secret key.
    The signature is the first SIGNATUREBYTES bytes of the return value.
    A copy of msg is in the remainder."""
    assert len(sk) == SECRETKEYBYTES
    sk = ctypes.create_string_buffer(sk, SECRETKEYBYTES) # const
    m = ctypes.create_string_buffer(msg, len(msg)) # const
    sig_and_msg = ctypes.create_string_buffer(len(msg) + SIGNATUREBYTES) # out
    sig_and_msg_len = ctypes.c_ulonglong() # out
    rc = _ed25519.crypto_sign(sig_and_msg,
                              ctypes.byref(sig_and_msg_len),
                              m, ctypes.c_ulonglong(len(m)),
                              sk)
    if rc != 0: # pragma no cover (no other return statement in C)
        raise ValueError("rc != 0", rc)
    return sig_and_msg.value[:sig_and_msg_len.value]


def crypto_sign_open(signed, vk):
    """Return message given signature+message and the verifying key."""
    assert len(vk) == PUBLICKEYBYTES
    sm = ctypes.create_string_buffer(signed, len(signed)) # const
    vk = ctypes.create_string_buffer(vk, len(vk)) # const
    newmsg = ctypes.create_string_buffer(len(signed)) # out
    newmsg_len = ctypes.c_ulonglong() # out
    rc = _ed25519.crypto_sign_open(newmsg, ctypes.byref(newmsg_len), 
                                   sm, len(sm), vk)
    if rc != 0:
        raise ValueError("rc != 0", rc)    
    return newmsg.value[:newmsg_len.value]

